webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/score/ScoreRepository.scala

```diff
@@ -0,0 +1,132 @@
+ * distributed under the license agreement of your choice.
+ * A "Related Module" means a set of sources files including their
+ * documentation that, without modification of the Source Code, enables
+ * supplementary functions or services in addition to those offered by
+ * the Software.
+ *
+ * Rudder is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+ *
+ *************************************************************************************
+ */
+
+package com.normation.rudder.score
+
+import com.normation.errors.IOResult
+import com.normation.inventory.domain.NodeId
+import com.normation.rudder.db.Doobie
+import doobie.Fragments
+import doobie.Get
+import doobie.Put
+import doobie.Read
+import doobie.Write
+import doobie.implicits._
+import doobie.implicits.toSqlInterpolator
+import zio.interop.catz._
+import zio.json.ast.Json
+
+trait ScoreRepository {
+
+  def getAll(): IOResult[Map[NodeId, List[Score]]]
+  def getScore(nodeId:    NodeId, name:  Option[String]): IOResult[List[Score]]
+  def getOneScore(nodeId: NodeId, name:  String):         IOResult[Score]
+  def saveScore(nodeId:   NodeId, score: Score):          IOResult[Unit]
+  def deleteScore(nodeId: NodeId, name:  Option[String]): IOResult[Unit]
+
+}
+
+class ScoreRepositoryImpl(doobie: Doobie) extends ScoreRepository {
+
+  import com.normation.rudder.db.json.implicits._
+  implicit val getScoreValue: Get[ScoreValue] = Get[String].temap(ScoreValue.fromString)
+  implicit val putScoreValue: Put[ScoreValue] = Put[String].contramap(_.value)
+
+  // implicit val stateWrite: Meta[Score] = new Meta(pgDecoderGet, pgEncoderPut)
+
+  implicit val scoreWrite: Write[(NodeId, Score)] = {
+    Write[(String, ScoreValue, String, String, Json)].contramap {
+      case (nodeId: NodeId, score: Score) =>
+        (nodeId.value, score.value, score.name, score.message, score.details)
+    }
+  }
+
+  implicit val scoreRead:       Read[Score]           = {
+    Read[(ScoreValue, String, String, Json)].map { d: (ScoreValue, String, String, Json) => Score(d._2, d._1, d._3, d._4) }
+  }
+  implicit val scoreWithIdRead: Read[(NodeId, Score)] = {
+    Read[(String, ScoreValue, String, String, Json)].map { d: (String, ScoreValue, String, String, Json) =>
+      (NodeId(d._1), Score(d._3, d._2, d._4, d._5))
+    }
+  }
+
+  import doobie._
+  override def getAll(): IOResult[Map[NodeId, List[Score]]] = {
+    val q = sql"select nodeId, score, name, message, details from scoreDetails "
+    transactIOResult(s"error when getting scores for node")(xa => q.query[(NodeId, Score)].to[List].transact(xa))
+      .map(_.groupMap(_._1)(_._2))
+  }
```

Comment:
Those tuple indices look really confusing between both implicits (but they look correct)
Normally doobie can generate the `Read[(NodeId, Score)]` by re-using the `Read[Score]` so we can remove the noise from `scoreWithIdRead` here